hrm-domain/scheduled-tasks/conversation-cleanup/index.ts

```diff
@@ -0,0 +1,234 @@
+        err,
+      );
+    }
+    ssmRetentionDays = MAX_CLEANUP_JOB_RETENTION_DAYS;
+  }
+
+  // For now we are limiting the retention days to 365 days for all jobs and allowing for a
+  // global override on a per account basis. This may need to epand to be more granular in the
+  // future.
+  return Math.min(MAX_CLEANUP_JOB_RETENTION_DAYS, ssmRetentionDays);
+};
+
+const PAGE_SIZE = 40;
+
+type ConversationListResponse = {
+  conversations: { sid: string }[];
+};
+
+const removeConversationsBySids = async (
+  accountSid: AccountSID,
+  conversationSids: string[],
+  attemptToClose: boolean,
+) => {
+  const twilioClient = await getClient({ accountSid });
+  // eslint-disable-next-line @typescript-eslint/no-loop-func
+  const removePromises = conversationSids.map(async sid => {
+    if (attemptToClose) {
+      console.debug(`Closing conversation ${sid} from account ${accountSid}`);
+      try {
+        const result = await twilioClient.conversations.v1.conversations
+          .get(sid)
+          .update({ state: 'closed' });
+        if (result.state === 'closed') {
+          console.debug(`Closed conversation ${sid} from account ${accountSid}`);
+          return;
+        }
+      } catch (e) {
+        console.warn(
+          `Error occurred while closing conversation ${sid} from account ${accountSid}, removing instead`,
+          e,
+        );
+      }
+    }
+    console.debug(`Removing conversation ${sid} from account ${accountSid}`);
+    const result = await twilioClient.conversations.v1.conversations.get(sid).remove();
+    if (result) {
+      console.debug(`Removed conversation ${sid} from account ${accountSid}`);
+    } else {
+      console.warn(`Failed to remove conversation ${sid} from account ${accountSid}`);
+    }
+  });
+  return Promise.all(removePromises);
+};
+const delay = (ms: number) => new Promise(res => setTimeout(res, ms));
+
+const ITERATION_DELAY = 2000;
+const MAX_ITERATIONS = 1000;
+const removeOldConversations = async (
+  accountSid: AccountSID,
+  retentionDays: number,
+  errorIfFound: boolean = false,
+  state?: string,
+) => {
+  const endDate = subDays(new Date(), retentionDays);
+  const authToken = await getSsmParameter(`/${hrmEnv}/twilio/${accountSid}/auth_token`);
+  let conversationsRetrieved = PAGE_SIZE;
+  let iteration = 0;
+  // Whilst we can filter conversations using the API, we can just repeatedly pull the first page.
+  // This is because everything we pull will no longer be valid for the query after we process it, so repeating the query will get new result set
+  // If we introduce filtering that needs to be done on the client, we will need to start paging through results.
+  // This is because otherwise we could request 100 results that all get filtered on the client, but repeating the same query will just keep getting those same 100 results back.
+  while (conversationsRetrieved >= PAGE_SIZE && iteration++ < 100) {
```

Comment:
```suggestion
  for (iteration = 0; conversationsRetrieved >= PAGE_SIZE && iteration < 100; iteration++) {
```
is easier to read in my opinion. I'm actually not even sure if iteration is `0` or `1` in the first iteration with the `while` xD. If `while` is preferred maybe move the increment at the end of each iteration rather than in the condition?

Reply:
I disagree about using a for loop being more readable. The max iterations is a 'safety valve', it's never intended to be hit unless something is broken. To me a `for()` loop with that many iterations looks like it's supposed to iterate 100 times during normal operation.

And the precise value doesn't really matter. Do we care if we stop at 99, 100, 101 iterations? It's just an arbitrary 'high number' to establish that we're possibly in an infinite loop

I also disagree that the increment is more readable buried in the loop body rather than right there where we check it, As is you can see 'it can only loop 100 times' in the condition, rather than having to look in 2 places to see that complete logic. If you really don't like the inline increment I can put it on the next line, it's not lost there.

But it should check against the constant rather than a magic number


```suggestion
  while (conversationsRetrieved >= PAGE_SIZE && iteration < MAX_ITERATIONS) {
    iteration++;
```